### DISCUSSION TITLE
Newbie - please help with Tutorial 1 - Cannot get POST to work...

### DISCUSSION THREAD

**Discussion by tseyfarth** [PostgREST/postgrest#3157] [Q&A] [answered]
Hello all,

I am a newbie to PostgREST, Curl and JSON.  

This code fails with a 
```
curl http://localhost:3000/todos -X POST \  -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.84QBxKjMKX676kcViolEUDew6UM61CYHmFYQmouCDxw"; \ -H "Content-Type: application/json" \  -d '{"task":"learn how to auth"}'
```

I get the following result:
```
{"code":"PGRST301","details":null,"hint":null,"message":"JWSError (JSONDecodeError \"Not valid base64url\")"}curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
```

The tutorial notes.config is:
```
db-uri = "postgres://authenticator:26455462@localhost:5432/axcysconnex"
#db-uri = "postgres://authenticator:eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kbyB1c2VyIn0.ndQRXT2TUeMXAbP2o5CzCUQpzgDV-dCqdZaTGQBVQdY@localhost:5432/axcysconnex"
db-schemas = "api"
db-anon-role = "web_anon"
jwt-secret           = "reallyreallyreallyreallyverysafe"
jwt-secret-is-base64 = false
server-port = 3000
```

Can anyone point me to where I have got it wrong?  I've tried so many iterations of this and still get a failure!

Also, the Tutorial 1  Shows this:
```
export TOKEN="<paste token here>"

curl http://localhost:3000/todos -X POST \
     -H "Authorization: Bearer $TOKEN"   \
     -H "Content-Type: application/json" \
     -d '{"task": "learn how to auth"}'
```

Using Windows command prompt, if I include the first line (with the token) windows tells me it is not a valid command.  How or where, can I set the token so it does not have to be included everytime, and just be $TOKEN??

Thank you,
Tim

**Comment by rotty3000** [comment]
There are 3 separate authorization bits that I think you might have gotten mixed up on:
1. the password for the database (e.g. [here](https://postgrest.org/en/stable/tutorials/tut0.html#step-2-install-postgresql))
2. the password for the *PostgREST* process to connect to the database = the `autenticator` account password (e.g. [bottom of this section here](https://postgrest.org/en/stable/tutorials/tut0.html#step-4-create-database-for-api))
3. the _secret_ used to mint (HS256) JWT tokens which is stored in the property/environment variable `jwt-secret/PGRST_JWT_SECRET` as described [here](https://postgrest.org/en/stable/tutorials/tut1.html#step-2-make-a-secret))

Consider in your example above; you placed the JWT token in the DB connect string as the password for the `authenticator` account to connect with the DB which is not quite right. That token should only be used in the curl commands. :)

Don't worry! I'm also a newbie to this project (started last week) and stumbled a few times doing the setup. What I would suggest is going back again and re-reading carefully because it does work as written. But if you continue to have issues, check out the gitter channel linked on the README OR visit the discussion form linked to this project.

**Reply by tseyfarth** [reply]
Hi rotty3000

Thank you for your reply.  I will try the discussion forum as well.

Since you are fresh on this, I thought I'd as a couple more things:

** I am not using Docker.

Postgrest is able to connect to the database - Tutorial0.  So the config file appears to be OK
```
C:\postgrest>postgrest tutorial.conf
15/Jan/2024:13:19:07 -0700: Starting PostgREST 12.0.2 (a4e00ff)...
15/Jan/2024:13:19:07 -0700: Attempting to connect to the database...
15/Jan/2024:13:19:07 -0700: Connection successful
15/Jan/2024:13:19:07 -0700: Listening on port 3000
15/Jan/2024:13:19:07 -0700: Config reloaded
15/Jan/2024:13:19:07 -0700: Listening for notifications on the pgrst channel
15/Jan/2024:13:19:07 -0700: Schema cache loaded
```

Using the command 'curl http://localhost:3000/todos' gives proper feedback:
```
C:\postgrest>curl http://localhost:3000/todos
[{"id":1,"done":false,"task":"finish tutorial 0","due":null},
 {"id":2,"done":false,"task":"pat self on back","due":null}]
```

So, I can access the schema/table that the tutorial creates.

This is the expanded .conf file where the jst-secret is added:

```
db-uri = "postgres://authenticator:26455462@localhost:5432/axcysconnex"
db-schemas = "api"
db-anon-role = "web_anon"
# The secret to verify the JWT for authenticated requests with.
# Needs to be 32 characters minimum.
jwt-secret = "reallyreallyreallyreallyverysafe"
jwt-secret-is-base64 = false
server-port = 3000
```

When moving to Tutorial 1, that's where the trouble starts.
**Again, not using docker **

This part of Tut1 is slightly confusing:
It says to "check the tutorial.conf (created in the previous tutorial) the line with the password:"

```
# PASSWORD MUST BE AT LEAST 32 CHARS LONG
cat tutorial.conf
```

This I do show in the above .conf file.  It is plain text and not the token.  The token was created using the plain text jwt-secret as the source per Tutorial 1.

In Step 4 of Tutorial 1, it says to:
"export TOKEN="<paste token here>"

When I try to run that command in Windows Cmd Promt, it is rejected by Windows.  Since I could not perform this line task, I put the token in the command

Original command from Tutorial 1:
```
curl http://localhost:3000/todos -X POST \
     -H "Authorization: Bearer $TOKEN"   \
     -H "Content-Type: application/json" \
     -d '{"task": "learn how to auth"}' 
```

Token	 
```
eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.SDjBjuJJxqjEfS84-pyesq5IRdLSzMbAJRbmJ79MOZ8
``` 

New Command	

```
 curl http://localhost:3000/todos -X POST \ 
     -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.SDjBjuJJxqjEfS84-pyesq5IRdLSzMbAJRbmJ79MOZ8"   \
     -H "Content-Type: application/json" \
     -d '{"task": "learn how to auth"}'
``` 

Command as one line -this is what was used:
```
curl http://localhost:3000/todos -X POST \      -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.SDjBjuJJxqjEfS84-pyesq5IRdLSzMbAJRbmJ79MOZ8"   \      -H "Content-Type: application/json" \      -d '{"task": "learn how to auth"}'
``` 


I get this error:
```
C:\postgrest>curl http://localhost:3000/todos -X POST \  -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.SDjBjuJJxqjEfS84-pyesq5IRdLSzMbAJRbmJ79MOZ8"; \ -H "Content-Type: application/json" \  -d '{"task":"learn how to auth"}'
{"code":"PGRST301","details":null,"hint":null,"message":"JWSError (JSONDecodeError \"Not valid base64url\")"}curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
``` 

I *think* the more specific question is, why am I getting this reply?
"JWSError (JSONDecodeError \"Not valid base64url\")"}curl: (3) URL rejected: Bad hostname

In this example, can you see where/why the URL is rejected?  Localhost is correct. The port 3000 is correct - so do not understand whats going on there.

One more question:
Is there an underscore or a dash in the command  jwt-secret = "....."

Thank you again for your reply and help!

Tim

**Comment by rotty3000** [comment]
> **Again, not using docker **

this is irrelevant because this only changes whether you use a config file or environment variables :)

> Postgrest is able to connect to the database

Great!

> It says to "check the tutorial.conf ...

Yeah, the section is titled [**Step 2. Make a Secret**](https://postgrest.org/en/stable/tutorials/tut1.html#step-2-make-a-secret) and then it uses 3 different terms to refer to exactly the same thing. Re-read that section and note that all these terms refer to the same thing; **secret** = **password** = **jwt-secret**.

> `jwt-secret-is-base64 = false`

Don't set this! This is probably why you're getting `Not valid base64url`

> `export TOKEN...`

This is a Unix/Linux shell thing. In windows I think it's more like `set TOKEN=...`. But as you discovered, just passing the token inline with the curl command should work too. In windows you might have done:

```shell
set TOKEN=....
curl ... -H "Authorization: Bearer %TOKEN%" ...
```

**Reply by wolfgangwalther** [reply]
> Yeah, the section is titled [Step 2. Make a Secret](https://postgrest.org/en/stable/tutorials/tut1.html#step-2-make-a-secret) and then it uses 3 different terms to refer to exactly the same thing. Re-read that section and note that all these terms refer to the same thing; secret = password = jwt-secret.

Ouch. Thanks for the hint. I created an [issue](https://github.com/PostgREST/postgrest-docs/issues/743) to rephrase that to use "secret" as the term throughout. That should hopefully cause less confusion.

**Comment by tseyfarth** [comment]
@rotty3000 
Thanks for the hint. I removed those items from the .conf file as you suggested.  That did fix the auth issue.  However, I still get an error woth the URL being bad:

From another response I put single quotes around the URL, but still get a bad URL result. 

```
C:\postgrest>curl 'localhost:3000/todos' -X POST \  -H "Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoidG9kb191c2VyIn0.84QBxKjMKX676kcViolEUDew6UM61CYHmFYQmouCDxw"; \ -H "Content-Type: application/json" \  -d '{"task":"learn how to auth"}'
curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
curl: (3) URL rejected: Bad hostname
```

Any other insights?  This should not be this hard!
Thank you,
Tim

**Reply by wolfgangwalther** [reply, marked as answer]
First of all, it's really hard to read the code snippets, commands and log outputs that you posted, because you were using single back-ticks only, which doesn't work well with multi-line strings. I changed all of those in the comments in this discussion to use triple back-ticks. Now I can actually read something.

The curl commands you are taking from the docs are written for linux. That means they have some specific syntax:
- Note the `\` at the end of each line in the curl command. Those are "line continuation" markers and allow to split a single command into multiple lines. In your case it seems like you are putting the whole command in a single line - so you must **remove the `\`**. A bit of googling makes me believe that the line continuation on the Windows command prompt can be done with `^` - so you could also replace it with that and keep the multiple lines.
- You added a `;` after the token. This is not there in the tutorial - where does it come from? You should remove that.

The errors you are seeing are telling you that curl is trying to understand `\` as a hostname.. which ofc does not work.

TLDR: Just remove the `\` and `;` and proceed happily.
